File: app/src/main/java/me/magnum/melonds/ui/SettingsActivity.java
package me.magnum.melonds.ui;

import android.os.Bundle;
import android.view.MenuItem;
import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.preference.ListPreference;
import androidx.preference.Preference;
import androidx.preference.PreferenceFragmentCompat;
import androidx.preference.PreferenceManager;
import com.github.angads25.filepicker.view.FilePickerPreference;
import me.magnum.melonds.R;
import me.magnum.melonds.utils.PreferenceDirectoryUtils;


public class SettingsActivity extends AppCompatActivity {
	private static Preference.OnPreferenceChangeListener sBindPreferenceSummaryToValueListener = new Preference.OnPreferenceChangeListener() {
		@Override
		public boolean onPreferenceChange(Preference preference, Object value) {
			String stringValue = value.toString();

			if (preference instanceof FilePickerPreference) {
				String directory = PreferenceDirectoryUtils.getSingleDirectoryFromPreference(stringValue);
				preference.setSummary(directory);
			} else if (preference instanceof ListPreference) {
				// For list preferences, look up the correct display value in
				// the preference's 'entries' list.
				ListPreference listPreference = (ListPreference) preference;
				int index = listPreference.findIndexOfValue(stringValue);

				// Set the summary to reflect the new value.
				preference.setSummary(
						index >= 0
								? listPreference.getEntries()[index]
								: null);

			} else {
				// For all other preferences, set the summary to the value's
				// simple string representation.
				preference.setSummary(stringValue);
			}
			return true;
		}
	};

	private static void bindPreferenceSummaryToValue(Preference preference) {
		// Set the listener to watch for value changes.
		preference.setOnPreferenceChangeListener(sBindPreferenceSummaryToValueListener);

		// Trigger the listener immediately with the preference's
		// current value.
		sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
				PreferenceManager
						.getDefaultSharedPreferences(preference.getContext())
						.getString(preference.getKey(), ""));
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setupActionBar();
		getSupportFragmentManager()
				.beginTransaction()
				.replace(android.R.id.content, new MainPreferencesFragment())
				.commit();
	}

	private void setupActionBar() {
		ActionBar actionBar = getSupportActionBar();
		if (actionBar != null) {
			// Show the Up button in the action bar.
			actionBar.setDisplayHomeAsUpEnabled(true);
		}
	}

	@Override
	public boolean onOptionsItemSelected(@NonNull MenuItem item) {
		int id = item.getItemId();
		if (id == android.R.id.home) {
			if (!super.onOptionsItemSelected(item)) {
				finish();
			}
			return true;
		}
		return super.onOptionsItemSelected(item);
	}

	public static class MainPreferencesFragment extends PreferenceFragmentCompat {
		@Override
		public void onCreatePreferences(Bundle savedInstanceState, String rootKey) {
			setPreferencesFromResource(R.xml.pref_main, rootKey);

			bindPreferenceSummaryToValue(findPreference("bios_dir"));
			bindPreferenceSummaryToValue(findPreference("sram_dir"));
		}
	}
}

File: app/src/main/java/me/magnum/melonds/ui/SettingsActivity.kt
package me.magnum.melonds.ui

import android.net.Uri
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AppCompatActivity
import androidx.preference.ListPreference
import androidx.preference.Preference
import androidx.preference.PreferenceFragmentCompat
import androidx.preference.PreferenceManager
import com.github.angads25.filepicker.view.FilePickerPreference
import me.magnum.melonds.R
import me.magnum.melonds.preferences.DirectoryPickerPreference
import me.magnum.melonds.utils.DirectoryPickerContract
import me.magnum.melonds.utils.FileUtils
import me.magnum.melonds.utils.PreferenceDirectoryUtils.getMultipleDirectoryFromPreference

class SettingsActivity : AppCompatActivity() {
    companion object {
        private val sBindPreferenceSummaryToValueListener = Preference.OnPreferenceChangeListener { preference, value ->
            when (preference) {
                is FilePickerPreference -> {
                    var directory = getMultipleDirectoryFromPreference(value.toString())
                    if (directory.isEmpty())
                        directory = arrayOf(preference.getContext().getString(R.string.not_set))

                    preference.setSummary(directory.joinToString("\n"))
                }
                is ListPreference -> {
                    // For list preferences, look up the correct display value in
                    // the preference's 'entries' list.
                    val index = preference.findIndexOfValue(value.toString())

                    // Set the summary to reflect the new value.
                    val summary = if (index >= 0)
                        preference.entries[index]
                    else
                        preference.getContext().getString(R.string.not_set)
                    preference.setSummary(summary)
                }
                is DirectoryPickerPreference -> {
                    if (value == null || value !is Set<*> || value.isEmpty())
                        preference.summary = preference.getContext().getString(R.string.not_set)
                    else {
                        val uris = value.mapNotNull { FileUtils.getAbsolutePathFromSAFUri(preference.context, Uri.parse(it as String)) }
                        preference.summary = uris.joinToString("\n")
                    }
                }
                else -> {
                    // For all other preferences, set the summary to the value's
                    // simple string representation.
                    preference.summary = value.toString()
                }
            }
            true
        }

        private fun bindPreferenceSummaryToValue(preference: Preference?) {
            if (preference == null)
                return

            // Set the listener to watch for value changes.
            preference.onPreferenceChangeListener = sBindPreferenceSummaryToValueListener

            // Trigger the listener immediately with the preference's
            // current value. Special handling for directory pickers since sets can't be converted
            // to string
            if (preference is DirectoryPickerPreference) {
                sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                        PreferenceManager
                                .getDefaultSharedPreferences(preference.context)
                                .getStringSet(preference.key, null))
            } else {
                sBindPreferenceSummaryToValueListener.onPreferenceChange(preference,
                        PreferenceManager
                                .getDefaultSharedPreferences(preference.context)
                                .getString(preference.key, null))
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupActionBar()
        supportFragmentManager
                .beginTransaction()
                .replace(android.R.id.content, MainPreferencesFragment())
                .commit()
    }

    private fun setupActionBar() {
        val actionBar = supportActionBar
        actionBar?.setDisplayHomeAsUpEnabled(true)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        if (id == android.R.id.home) {
            if (!super.onOptionsItemSelected(item)) {
                finish()
            }
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    class MainPreferencesFragment : PreferenceFragmentCompat() {
        override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
            setPreferencesFromResource(R.xml.pref_main, rootKey)
            val romDirsPreference = findPreference<DirectoryPickerPreference>("rom_search_dirs")!!
            val biosDirPreference = findPreference<DirectoryPickerPreference>("bios_dir")!!
            val sramDirPreference = findPreference<DirectoryPickerPreference>("sram_dir")!!
            bindPreferenceSummaryToValue(romDirsPreference)
            bindPreferenceSummaryToValue(biosDirPreference)
            bindPreferenceSummaryToValue(sramDirPreference)

            val romPickerLauncher = registerForActivityResult(DirectoryPickerContract(), romDirsPreference::onDirectoryPicked)
            val biosPickerLauncher = registerForActivityResult(DirectoryPickerContract(), biosDirPreference::onDirectoryPicked)
            val sramPickerLauncher = registerForActivityResult(DirectoryPickerContract(), sramDirPreference::onDirectoryPicked)

            romDirsPreference.setOnPreferenceClickListener {
                romPickerLauncher.launch(null)
                true
            }
            biosDirPreference.setOnPreferenceClickListener {
                biosPickerLauncher.launch(null)
                true
            }
            sramDirPreference.setOnPreferenceClickListener {
                sramPickerLauncher.launch(null)
                true
            }
        }
    }
}